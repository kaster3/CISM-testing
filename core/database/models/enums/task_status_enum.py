from enum import Enum


class TaskStatusEnum(Enum):
    NEW = "Новая задача"
    IN_PROGRESS = "В процессе работы"
    COMPLETED = "Завершено успешно"
    ERROR = "Ошибка"
