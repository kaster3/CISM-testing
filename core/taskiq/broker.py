import os

from taskiq import AsyncBroker, InMemoryBroker
from taskiq_aio_pika import AioPikaBroker

from core.settings import settings

env = os.environ.get("ENVIRONMENT")

broker: AsyncBroker = AioPikaBroker(
    url=settings.rabbitmq.url,
)

if env and env == "pytest":
    broker = InMemoryBroker()
