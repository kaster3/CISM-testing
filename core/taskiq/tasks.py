import asyncio
import logging
import random
from typing import Annotated

from taskiq import TaskiqDepends

from api.api_v1.task.dependencies import get_task_repository
from api.api_v1.task.repository import TaskRepository
from api.api_v1.task.schemas import TaskUpdatePartial
from core.database.models.enums.task_status_enum import TaskStatusEnum
from .broker import broker


log = logging.getLogger(__name__)


@broker.task
async def task_process(
        task_in: int,
        repo: Annotated[TaskRepository, TaskiqDepends(get_task_repository)],
) -> None:
    log.info("Starting task process")
    task = await repo.get_task(task_id=task_in)
    task = await repo.update_task(
        task=task, task_in=TaskUpdatePartial(status=TaskStatusEnum.IN_PROGRESS)
    )
    log.info("Updated task with status: %s" % task.status.value)
    await asyncio.sleep(random.randint(5, 10))
    finished_status = TaskStatusEnum.COMPLETED if random.random() < 0.5 else TaskStatusEnum.ERROR
    task = await repo.update_task(task=task, task_in=TaskUpdatePartial(status=finished_status))
    log.info("Completed task with status: %s" % task.status.value)
