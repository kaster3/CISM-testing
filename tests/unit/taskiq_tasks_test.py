import pytest
from pytest_mock import MockerFixture

from core.database.models.enums.task_status_enum import TaskStatusEnum
from core.taskiq.tasks import TaskUpdatePartial, task_process


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "random_value, expected_status",
    [
        (0.4, TaskStatusEnum.COMPLETED),
        (0.6, TaskStatusEnum.ERROR),
    ],
)
async def test_task_process(
    mocker: MockerFixture,
    random_value: float,
    expected_status: TaskStatusEnum,
    task_id: int = 1,
):
    mock_repo = mocker.AsyncMock()

    mock_task = mocker.AsyncMock()
    mock_task.id = 1
    mock_task.name = "Test Task"
    mock_task.status = TaskStatusEnum.NEW

    mock_repo.get_task.return_value = mock_task
    mock_repo.update_task.return_value = mock_task

    mocker.patch("asyncio.sleep", return_value=None)
    mocker.patch("random.random", return_value=random_value)

    await task_process(task_in=task_id, repo=mock_repo)

    mock_repo.get_task.assert_called_once_with(task_id=task_id)

    expected_calls = [
        mocker.call(task=mock_task, task_in=TaskUpdatePartial(status=TaskStatusEnum.IN_PROGRESS)),
        mocker.call(task=mock_task, task_in=TaskUpdatePartial(status=expected_status)),
    ]
    mock_repo.update_task.assert_has_calls(expected_calls, any_order=False)
