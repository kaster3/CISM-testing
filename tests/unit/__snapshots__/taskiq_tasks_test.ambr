# serializer version: 1
# name: test_task_process[0.4-TaskStatusEnum.COMPLETED]
  TaskCreate(name='Test Task')
# ---
# name: test_task_process[0.6-TaskStatusEnum.ERROR]
  TaskCreate(name='Test Task')
# ---
